src/CodeGen_RS_Dev.cpp

```diff
@@ -0,0 +1,552 @@
+
+    // Add Renderscript standard set of metadata.
+    NamedMDNode *rs_export_foreach_name =
+        module->getOrInsertNamedMetadata("#rs_export_foreach_name");
+    rs_export_foreach_name->addOperand(MDNode::get(
+        *context,
+        vec<LLVMMDNodeArgumentType>(MDString::get(*context, "root"))));
+    rs_export_foreach_name->addOperand(MDNode::get(
+        *context,
+        vec<LLVMMDNodeArgumentType>(MDString::get(*context, kernel_name))));
+
+    const char* kernel_signature = "57";
+    NamedMDNode *rs_export_foreach =
+        module->getOrInsertNamedMetadata("#rs_export_foreach");
+    rs_export_foreach->addOperand(MDNode::get(
+        *context, vec<LLVMMDNodeArgumentType>(MDString::get(*context, "0"))));
+    rs_export_foreach->addOperand(MDNode::get(
+        *context, vec<LLVMMDNodeArgumentType>(MDString::get(
+                      *context, kernel_signature))));
+
+    // Now verify the function is ok
+    verifyFunction(*function);
+
+    // Finally, verify the module is ok
+    verifyModule(*module);
+
+    debug(2) << "Done generating llvm bitcode for RS\n";
+
+    // Clear the symbol table
+    for (size_t i = 0; i < arg_sym_names.size(); i++) {
+        sym_pop(arg_sym_names[i]);
+    }
+}
+
+void CodeGen_RS_Dev::init_module() {
+    debug(2) << "CodeGen_RS_Dev::init_module\n";
+    init_context();
+#ifdef WITH_RS
+    delete module;
+    module = get_initial_module_for_rs_device(target, context);
+#endif
+}
+
+llvm::Triple CodeGen_LLVM::get_target_triple() const {
+    return Triple(Triple::normalize("armv7-none-linux-gnueabi"));
+}
+
+llvm::DataLayout CodeGen_RS_Dev::get_data_layout() const {
+    return llvm::DataLayout("e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64");
+}
+
+//
+// Loops become kernels. There should be no explicit loops in
+// generated RenderScript code.
+//
+void CodeGen_RS_Dev::visit(const For *loop) {
+    debug(2) << "RS: Visiting for loop, loop->name is " << loop->name
+             << " is_gpu_var? " << is_gpu_var(loop->name) << "\n";
+    if (is_gpu_var(loop->name)) {
+        // Whether it's thread-parallelization loop or loop over
+        // coordinate variables, collapse them going straight to the body
+        // because Renderscript takes care of setting up the loops.
+        // We just need to produce a kernel code.
+        loop->body.accept(this);
+    } else {
+        user_assert(loop->for_type != ForType::Parallel)
+            << "Cannot use loops inside RS kernel\n";
+        CodeGen_LLVM::visit(loop);
+    }
+}
+
+void CodeGen_RS_Dev::visit(const Allocate *alloc) {
+    debug(2) << "RS: Allocate " << alloc->name << " on device\n";
+    codegen(alloc->body);
+}
+
+void CodeGen_RS_Dev::visit(const Free *f) {
+    // TODO(aam): Implement this.
+    debug(2) << "RS: Free on device\n";
+}
+
+llvm::Function *CodeGen_RS_Dev::fetch_GetElement_func(int width) {
+    std::string func_name = width == 1
+                                ? "_Z20rsGetElementAt_uchar13rs_allocationjjj"
```

Comment:
As far as I understand these are unlikely to break as they correspond to RS publicly available API calls. We hardcode them here so that the code running on the device can make those RS-specific calls to get/update buffer elements.